src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs

```diff
@@ -644,26 +646,37 @@ private void Dispose()
+                        // one of two places, and if we don't find it we generate it from the type (MakeImplicitTransforms)
                         Type argType = args.GetType();
-                        if (_expectedArgType != argType)
+                        TransformSpec implicitTransforms;
+                        if (argType == _firstImplicitFieldsType && _firstImplicitFieldsTypeTransforms != null)
                         {
-                            // Figure out the default properties to send on to EventSource.  These are all string or primitive properties.  
-                            _implicitTransforms = null;
-                            TransformSpec newSerializableArgs = null;
-                            TypeInfo curTypeInfo = argType.GetTypeInfo();
-                            foreach (var property in curTypeInfo.DeclaredProperties)
-                            {
-                                var propertyType = property.PropertyType;
-                                newSerializableArgs = new TransformSpec(property.Name, 0, property.Name.Length, newSerializableArgs);
-                            }
-                            _expectedArgType = argType;
-                            _implicitTransforms = Reverse(newSerializableArgs);
+                            // This is the hoped for case.  We hit the singleton cache that we have already set.  
+                            implicitTransforms = _firstImplicitFieldsTypeTransforms;
+                        }
+                        else if (_firstImplicitFieldsType == null)
+                        {
+                            // The singleton cache slot is empty, so create the implicit transforms for that type and cache it.  
+                            implicitTransforms = MakeImplicitTransforms(argType);
+                            // Threads race to claim this slot but only one wins, and gets the right to update _firstImplicitFieldsTypeTransforms. 
+                            // We never modify it after that.   
+                            if (implicitTransforms != null && Interlocked.CompareExchange(ref _firstImplicitFieldsType, argType, null) == null)
+                                _firstImplicitFieldsTypeTransforms = implicitTransforms;
+                        }
+                        else
+                        {
+                            // This should only happen when you are wildcarding your events (reasonably rare).   In that case you will probably need many types
+                            // Note currently we don't limit the cache size, but it is limtied by the number of distinct types of objects passed to DiagnosticSource.Write.  
+                            if (_implicitTransformsTable == null)
+                                _implicitTransformsTable = new ConcurrentDictionary<Type, TransformSpec>(1, 8);
+                            implicitTransforms = _implicitTransformsTable.GetOrAdd(argType, MakeImplicitTransforms);
```

Comment:
What happens if two threads are racing here?  Couldn't they both end up initializing _implicitTransformsTable, adding to their own copy, such that what's actually published is incomplete and may not contain what was just added from one thread's perspective?

Reply:
Yes, there was a race but I didn't care since the cache does not affect semantics.   But I will grant you that it is less error prone the fewer races we tolerate, so I am OK putting in a CompareExchange instead of an assignment.